proposals/csharp-7.0/tuples.md

```diff
@@ -0,0 +1,437 @@
+(string x, byte y, var z) = (null, 1, 2);
+(string x, byte y) t = (null, 1);
+```
+
+Or for deconstructing assignment:
+
+```csharp
+string x;
+byte y;
+
+(x, y) = (null, 1);
+(x, y) = (y, x); // swap!
+```
+
+The evaluation order of deconstruction assignment expressions is "breadth first":
+
+1. Evaluate the LHS: Evaluate each of the expressions inside of it one by one, left to right, to yield side effects and establish a storage location for each.
+1. Evaluate the RHS: Evaluate each of the expressions inside of it one by one, left to right to yield side effects
+1. Convert each of the RHS expressions to the LHS types expected, one by one, left to right.
+1. Assign each of the conversion results from 3 to the storage locations found in 
+
+> **Note to reviewers**: I found this in the LDM notes for July 13-16, 2016. I don't think it is still accurate:
+
+A deconstructing assignment is a *statement-expression* whose type could be `void`.
+
+### Duality with underlying type
+
+Tuples map to underlying types of particular names.
+
+```csharp
+System.ValueTuple<T1, T2>
+System.ValueTuple<T1, T2, T3>
+...
+System.ValueTuple<T1, T2, T3,..., T7, TRest>
+```
+
+Tuple types behave exactly like underlying types. The only additional enhancement is the more expressive field names given by the programmer.
+
+```csharp
+var t = (sum: 0, count: 1);
+t.sum   = 1;        // sum   is the name for the field #1
+t.Item1 = 1;        // Item1 is the name of underlying field #1 and is also available
+
+var t1 = (0, 1); // tuple omits the field names.
+t.Item1 = 1;     // underlying field name is still available
+
+t.ToString();     // ToString on the underlying tuple type is called.
+
+System.ValueTuple<int, int> vt = t; // identity conversion
+(int moo, int boo) t2 = vt;         // identity conversion
+
+```
+
+Because of the dual nature of tuples, it is not allowed to assign field names that overlap with preexisting member names of the underlying type. The only exception is the use of predefined `Item1`, `Item2`,...`ItemN` at corresponding position N, since that would not be ambiguous.
+
+```csharp
+var t =  (ToString: 0, ObjectEquals: 1);  // error: names match underlying member names
+var t1 = (Item1: 0, Item2: 1);            // valid
+var t2 = (misc: 0, Item1: 1);             // error: "Item1" was used in a wrong position
+```
+
+If the tuple is bigger than the limit of 7, the implementation will nest the "tail" as a tuple into the eighth element recursively. This nesting is visible by accessing the `Rest` field of a tuple, but that field my be hidden from e.g. auto-completion, just as the `ItemX` field names may be hidden but allowed when a tuple has named elements.
+
+A well formed "big tuple" will have names `Item1` etc. all the way up to the number of tuple elements, even though the underlying type doesn't physically have those fields directly defined. The same goes for the tuple returned from the `Rest` field, only with the numbers "shifted" appropriately.
```

Comment:
>"big tuple" [](start = 14, length = 11)

define

Reply:
Rex has re-written this paragraph. The "big tuple" term no longer appears.